app/actions/user.ts

```diff
@@ -1,171 +1,153 @@
-import { providers } from "ethers";
+import { providers, ethers } from "ethers";
 import { Thunk } from "state";
 
-import { addresses } from "@klimadao/lib/constants";
 import {
   formatUnits,
-  trimStringDecimals,
   getContract,
+  createContractsObject,
+  getAllowance,
+  getTokensFromSpender,
+  getTokenDecimals,
 } from "@klimadao/lib/utils";
-import {
-  setBalance,
-  setExerciseAllowance,
-  setStakeAllowance,
-  setWrapAllowance,
-  setDomains,
-} from "state/user";
+import { AllowancesFormatted } from "@klimadao/lib/types/allowances";
+import { setBalance, updateAllowances, setDomains } from "state/user";
 
 import { getKns, getEns } from "./utils";
 
+const assets = [
+  "bct",
+  "nct",
+  "mco2",
+  "ubo",
+  "nbo",
+  "usdc",
+  "klima",
+  "sklima",
+  "wsklima",
+  "pklima",
+] as const;
+type Asset = typeof assets[number];
+
+const spenders = [
+  "staking_helper",
+  "staking",
+  "wsklima",
+  "pklima_exercise",
+] as const;
+
+type TokenValueFormatted = {
+  [K in Asset]: string;
+};
+
+const getBalance = async (params: {
```

Comment:
again i still think its better to do a formatted string rather than an object. the outer fn should always be able to use array index or the `token` string itself to know which value this is

if we return an object we might end up doing a lot of merge boiler plate if we reuse this in a lot of places for simple balance queries

Reply:
Done